azurerm/monitor_diagnostics.go

```diff
@@ -0,0 +1,48 @@
+package azurerm
+
+import "strings"
+
+type MonitorDiagnosticId struct {
+	ResourceID string
+	Name       string
+}
+
+func parseMonitorDiagnosticId(monitorId string) MonitorDiagnosticId {
+	var returnId MonitorDiagnosticId
+	returnId.ResourceID = retrieveResourceId(monitorId)
+	returnId.Name = retrieveDiagnosticName(monitorId)
+	return returnId
+}
+
+func retrieveResourceId(diagnosticSettingId string) string {
+	substring := "/providers/microsoft.insights/diagnosticSettings/"
+	return diagnosticSettingId[0:strings.Index(diagnosticSettingId, substring)]
+}
+
+func retrieveDiagnosticName(diagnosticSettingId string) string {
+	substring := "/providers/microsoft.insights/diagnosticSettings/"
+	return diagnosticSettingId[len(substring)+strings.Index(diagnosticSettingId, substring):]
+}
+
+func getAllDiagnosticSettings(targetResourceId string, meta interface{}) (*[]interface{}, *[]interface{}, error) {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsCategoryClient
+	ctx := meta.(*ArmClient).StopContext
+	returnMetricSettings := make([]interface{}, 0)
+	returnLogSettings := make([]interface{}, 0)
+
+	categoryList, err := client.List(ctx, targetResourceId)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, item := range *categoryList.Value {
+		if item.DiagnosticSettingsCategory.CategoryType == "Metrics" {
+			returnMetricSettings = append(returnMetricSettings, *item.Name)
+		}
+		if item.DiagnosticSettingsCategory.CategoryType == "Logs" {
```

Comment:
after pulling down the code and looking into this, I think this belongs as a data source; something like:

```
data "azurerm_diagnostic_setting_categories" "test" {
  target_resource_id = "/path/to/resource"
}

outputs "log_names" {
  value = "${data.azurerm_diagnostic_setting_categories.test.log_names}" # or log_types, or something?
}

outputs "metric_names" {
  value = "${data.azurerm_diagnostic_setting_categories.test.metric_names}" # or metric_types, or something?
}
```

these could then be used [with some of the changes coming in Terraform 0.12 in the near future](https://www.hashicorp.com/blog/hashicorp-terraform-0-12-preview-for-and-for-each) to do something like:

```
resource "azurerm_diagnostics_setting" "example" {
  # ...

  dynamic "metric_rule" {
    for_each = data.azurerm_diagnostic_setting_categories.test.metric_names

    content {
      name = metric.name
      # ...
    }
  }
}
```